hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/SCMNodeManager.java

```diff
@@ -400,6 +405,66 @@ public void processNodeReport(DatanodeDetails datanodeDetails,
     }
   }
 
+  /**
+   * Process Layout Version report.
+   *
+   * @param datanodeDetails
+   * @param layoutVersionReport
+   */
+  @Override
+  public void processLayoutVersionReport(DatanodeDetails datanodeDetails,
+                                LayoutVersionProto layoutVersionReport) {
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Processing Layout Version report from [datanode={}]",
+          datanodeDetails.getHostName());
+    }
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("HB is received from [datanode={}]: <json>{}</json>",
+          datanodeDetails.getHostName(),
+          layoutVersionReport.toString().replaceAll("\n", "\\\\n"));
+    }
+
+    if (layoutVersionReport != null) {
+      int scmSlv = scmLayoutVersionManager.getSoftwareLayoutVersion();
+      int scmMlv = scmLayoutVersionManager.getMetadataLayoutVersion();
+      UpgradeFinalizer.Status scmUpgradeState =
+          scmLayoutVersionManager.getUpgradeState();
+      int dnSlv = layoutVersionReport.getSoftwareLayoutVersion();
+      int dnMlv = layoutVersionReport.getMetadataLayoutVersion();
+
+      // If the data node slv is > scm slv => log error condition
+      if (dnSlv > scmSlv) {
```

Comment:
Do we also need to add the condition (dnMLV > scmMLV) here (and at register step)?

Reply:
- yes, we need the additional check here. The nodes that are already registered they will send the heartbeat when they are booted.
- I will add an end to end test case for this once the workflow is complete.

Reply:
@prashantpogde End to End test cases make sense for integration or acceptance tests. Is it possible to add a unit test just for the SCMNodeManager.processLayoutVersion method? Since it has a few condition checks (path flows), it may be a good candidate for a unit test based on mocks.

Reply:
yup